pkg/platform/runtime/buildscript/file.go

```diff
@@ -22,40 +22,59 @@ type projecter interface {
 	Name() string
 }
 
-func NewScriptFromProject(proj projecter, auth *authentication.Auth) (*Script, error) {
-	return newScriptFromFile(filepath.Join(proj.ProjectDir(), constants.BuildScriptFileName), proj.Owner(), proj.Name(), auth)
-}
+var ErrBuildscriptNotExist = errors.New("Build script does not exist")
 
-func newScriptFromFile(path, org, project string, auth *authentication.Auth) (*Script, error) {
-	if data, err := fileutils.ReadFile(path); err == nil {
-		return NewScript(data)
-	} else if !errors.Is(err, os.ErrNotExist) {
-		return nil, errs.Wrap(err, "Could not read build script")
-	}
+// ScriptFromProjectWithFallback will source the buildscript from the project, and create it if it does not exist.
+func ScriptFromProjectWithFallback(proj projecter, auth *authentication.Auth) (*Script, error) {
+	path := filepath.Join(proj.ProjectDir(), constants.BuildScriptFileName)
 
-	logging.Debug("Build script does not exist. Creating one.")
-	commitId, err := localcommit.Get(filepath.Dir(path))
-	if err != nil {
-		return nil, errs.Wrap(err, "Unable to get the local commit ID")
-	}
-	buildplanner := model.NewBuildPlannerModel(auth)
-	expr, err := buildplanner.GetBuildExpression(commitId.String())
-	if err != nil {
-		return nil, errs.Wrap(err, "Unable to get the remote build expression")
-	}
-	script, err := NewScriptFromBuildExpression(expr)
+	buildscript, err := ScriptFromFile(path)
 	if err != nil {
-		return nil, errs.Wrap(err, "Unable to convert build expression to build script")
+		if !errors.Is(err, os.ErrNotExist) {
+			return nil, errs.Wrap(err, "Could not read build script from file")
+		}
+
+		logging.Debug("Build script does not exist. Creating one.")
+		commitId, err := localcommit.Get(filepath.Dir(path))
+		if err != nil {
+			return nil, errs.Wrap(err, "Unable to get the local commit ID")
+		}
+		buildplanner := model.NewBuildPlannerModel(auth)
+		expr, err := buildplanner.GetBuildExpression(commitId.String())
+		if err != nil {
+			return nil, errs.Wrap(err, "Unable to get the remote build expression")
+		}
+		buildscript, err = NewScriptFromBuildExpression(expr)
+		if err != nil {
+			return nil, errs.Wrap(err, "Unable to convert build expression to build script")
+		}
+		err = fileutils.WriteFile(path, []byte(buildscript.String()))
+		if err != nil {
+			return nil, errs.Wrap(err, "Unable to write build script")
+		}
 	}
-	err = fileutils.WriteFile(path, []byte(script.String()))
+
+	return buildscript, nil
+}
+
+func ScriptFromProject(proj projecter) (*Script, error) {
+	path := filepath.Join(proj.ProjectDir(), constants.BuildScriptFileName)
+	return ScriptFromFile(path)
+}
+
+func ScriptFromFile(path string) (*Script, error) {
+	data, err := fileutils.ReadFile(path)
 	if err != nil {
-		return nil, errs.Wrap(err, "Unable to write build script")
+		if errors.Is(err, os.ErrNotExist) {
+			return nil, errs.Pack(err, ErrBuildscriptNotExist)
```

Comment:
For my own information, why `errs.Pack()` instead of `errs.Wrap()`?

Reply:
Let's say I implement 

```
type ErrBuildscriptNotExist struct {
    error
}
```

I can't just instantiate that as `ErrBuildScriptNotExist{wrappedErr}`, because then the type of `wrappedErr` can no longer be interpreted. What I could do is add an `Unwrap()` method to  this type, which returned the wrapped error, but that's a lot of boilerplate.

So to save myself some boilerplate and still easily be able to communicate two types I use `errs.Pack()`. 

This does mean that `ErrBuildscriptNotExist` does not wrap its underlying error, instead it lives adjacent to it. To me the compromise is worth it. If we were to require types to always go to the extent of adding Unwrap logic they'd be painful to use and we'd try to avoid using them, which ultimately is worse than the caveats of `.Pack()`.